File: Data Structures & Algorithms/count-number-of-islands/submission-0.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        ROWS = len(grid)-1
        COLS = len(grid[0])-1

        
        directions = [[0,1],[1,0],[0,-1],[-1,0]]
        islands = 0
        def dfs(r,c):
            if r>ROWS or c >COLS or r<0 or c<0 or grid[r][c]=="0":
                return
            grid[r][c] = "0"

            for dr,dc in directions:
                dfs(r+dr,c+dc)
        for r in range(len(grid)):
            for c in range(len(grid[r])):
                if grid[r][c] == "1":
                    islands+=1
                    dfs(r,c)
        
        return islands
        
            
            

        

File: Data Structures & Algorithms/surrounded-regions/submission-0.java
class Solution {
    private int ROWS,COLS;
    public void solve(char[][] board) {
        ROWS = board.length;
        COLS = board[0].length;


        for(int i = 0 ; i <ROWS ; i++){
            if(board[i][0]=='O'){
                dfs(board,i,0);
            }
            if(board[i][COLS-1]=='O'){
                dfs(board,i,COLS-1);
            }
        }

        for(int j =0; j<COLS; j++){
            if(board[0][j]=='O'){
                dfs(board,0,j);
            }
            if(board[ROWS-1][j]=='O'){
                dfs(board,ROWS-1,j);
            }
        }

        for(int r = 0 ; r<ROWS ; r++){
            for(int c = 0 ; c < COLS ; c++){
              if(board[r][c]=='O'){
                board[r][c] = 'X';
              }else if(board[r][c]=='T'){
                board[r][c] = 'O';
              }    
            }
        }
        
    }

    public void dfs(char[][] board, int r ,int c){
        if(r<0 || c <0 || r==ROWS || c==COLS || board[r][c]!='O'){
            return;
        }

        board[r][c] = 'T';
        dfs(board,r,c+1);
        dfs(board,r,c-1);
        dfs(board,r+1,c);
        dfs(board,r-1,c);
    }
}
